### ISSUE TITLE
Custom eval metric

### ISSUE THREAD

**Issue by ecoskian** [mljar/mljar-supervised#390]
Hello,

I was wondering if it's possible to add fully custom eval metrics. 

My specific use case is one where I would like to add up values of an arbitrary vector for all predictions that exceed a (percentile) threshold. In general, however, it would be great to have the option to decouple the eval metric used for fitting from the one used for evaluation/tuning.

Ideally, the user would be able to supply a function such as those in sklearn.metrics, accepting target values and predictions and returning a float. Whether to assume minimisation or maximisation (or have an extra parameter) isn't particularly important, imho.

Thoughts?

**Comment by pplonski** [comment]
@ecoskian should be possible! Do you need this soon?

**Comment by ecoskian** [comment]
@pplonski sounds exciting! Ideally soon would be great. Looking at the results of the generated models for a problem I'm actively working on, I'd expect a major improvement. Happy to test things out, too 🙂

**Comment by pplonski** [comment]
@ecoskian used defined eval metric is available for testing! :) Changes are in the `dev` branch. 

The user-defined eval metric should always minimize. It should have an interface like:

```py
def my_custom_metric(y_true, y_predicted,sample_weight=None):
    # compute the score ...
    return score
```


To set your eval metric just pass a name of the function as `eval_metric` argument:

```py
automl = AutoML(eval_metric=my_custom_metric, ...)
automl.fit(X, y)
```

Eval metric will be used for early stopping and model selection in AutoML (in all modes, even in Optuna). 

To install package from the `dev` branch:
```
pip install -q -U git+https://github.com/mljar/mljar-supervised.git@dev
```

@ecoskian looking for your feedback. Thank you!

**Comment by ecoskian** [comment]
@pplonski thanks for this, seems to work perfectly!

**Comment by bridgeso** [comment]
Is it possible to set the metric for the AutoML to be Precision on only the True labels, rather than the average Precision in a Binary classification?
thanks

**Comment by pplonski** [comment]
You can pass to `eval_metric` method with your custom metric. It can be any metric defined by you.

Please let me know if it works for you.

**Comment by bridgeso** [comment]
> You can pass to `eval_metric` method with your custom metric. It can be any metric defined by you.
> 
> Please let me know if it works for you.

I ran into an error, I tried generating a new basic precision metric. 

<img width="570" alt="2021-09-19 (1)" src="https://user-images.githubusercontent.com/72578450/133937248-4c9b90c8-be69-43cb-87b6-37ae77da3bff.png">

<img width="669" alt="2021-09-19 (2)" src="https://user-images.githubusercontent.com/72578450/133937289-3b594567-90c8-4ef1-8cb0-e2d73f559ea4.png">

**Comment by pplonski** [comment]
You need to apply some threshold on predicted values. Target values are integers with classes. Predicted values are probability of class. You can apply 0.5 threshold for example.  You need to do it in your custom metric method.

**Comment by bridgeso** [comment]
> You need to apply some threshold on predicted values. Target values are integers with classes. Predicted values are probability of class. You can apply 0.5 threshold for example. You need to do it in your custom metric method.

So I've rounded the predicted values to the nearest whole number 0 or 1 and converted them to Boolean type but still getting an error and the model is not training. 

```
from sklearn.metrics import precision_score
def pos_precision(y_true, y_predicted,sample_weight=None):
  df = pd.DataFrame(y_predicted)
  df['target'] = round(df[0]).astype('bool')
  prec = precision_score(y_true, df['target'])
  return prec

automl_metric1 = AutoML(results_path="AutoML_metric_9", eval_metric = pos_precision, algorithms=['Decision Tree'])
automl_metric1.fit(train_x1, train_y1)  
```

>                    0  target
>0     0.012152   False
>1     0.012152   False
>2     0.952322   True

<img width="603" alt="2021-09-20" src="https://user-images.githubusercontent.com/72578450/133945779-33be8ef2-84a6-48e1-a152-1e49d66c5b74.png">

**Comment by pplonski** [comment]
@bridgeso sorry for vague comment. I was out of my desktop.

You can find some examples of metrics in MLJAR AutoML code: https://github.com/mljar/mljar-supervised/blob/master/supervised/utils/metric.py#L72

The important thing to remember is that custom metric needs to be minimized. In your case, you need to multiply by `-1`.

Your custom metric code should look like:

```py

from sklearn.metrics import precision_score
def pos_precision(y_true, y_predicted,sample_weight=None):
    # asure proper type
    if isinstance(y_true, pd.DataFrame):
        y_true = np.array(y_true)
    if isinstance(y_predicted, pd.DataFrame):
        y_predicted = np.array(y_predicted)

    if len(y_predicted.shape) == 2 and y_predicted.shape[1] == 1:
        y_predicted = y_predicted.ravel()

    # cast to integer
    if len(y_predicted.shape) == 1:
        y_predicted = (y_predicted > 0.5).astype(int)
    else:
        y_predicted = np.argmax(y_predicted, axis=1)

    val = precision_score(y_true, y_predicted, sample_weight=sample_weight)

    return -val
```

I tested the above code with your data and code and it works. Please remember that all metrics values will be displayed as negative.

**Comment by johngouf** [comment]
Hi, I bumped into this issue and was more than helpful to understand how to build a custom metric.
However I am looking to introduce a custom metric which needs the rest of the row of the dataframe as an input to do some lookups.
Would it be possible?

**Comment by pplonski** [comment]
Hi @johngouf,

It should be possible. You can have access to your additional data in custom metric:

```python
# load additional data
df_additional = pd.read_csv("additional_data.csv")

def my_custom_metric(y_true, y_predicted,sample_weight=None):
    # access additional data
    df_additional['predictions'] = y_predicted 
    df_additional['target'] = y_true

    # do needed operations

    # compute the score ...
    return score

# train AutoML
automl = AutoML(eval_metric=my_custom_metric, ...)
automl.fit(X, y)
```

Please let me know if it works for you.

**Comment by luxiaolei** [comment]
> 

Hi, may I ask does `y_true` carries index information?

**Comment by pplonski** [comment]
@luxiaolei no it doesn't. The index is reset before the training.

**Comment by johngouf** [comment]
Actually that was issue, that I need to somehow pass the index and lookup information for the specific row

**Comment by pplonski** [comment]
I created an issue for this, I would need to check how it can be implemented https://github.com/mljar/mljar-supervised/issues/497

**Comment by luxiaolei** [comment]
my situation is the same as @johngouf . Alternatively I am thinking of,  is that possible to get the model instance of each CV after `fit`,  so that I can do the custom metric calculation outside.

**Comment by pplonski** [comment]
@luxiaolei great idea! For each model, there are saved out-of-fold predictions (combined from all folds with original ordering) or validation predictions depending on the validation strategy. They have unchanged ordering from training data. They are saved in CSV format, should be possible to load them and compute custom metrics with index.

**Comment by luxiaolei** [comment]
@pplonski great, thanks:) however I couldn't find those csv files, I saw there are `folds/fold_{n}_train/validation_indices.npy` files

**Comment by pplonski** [comment]
Please check inside each model directory.

**Comment by luxiaolei** [comment]
> 

ah got it, thanks:)

**Comment by Serif-Kaya** [comment]
While trying to access additional data in my custom metric as you mentioned above I see this error

`Failed to optimize method "evaluate" in the passed object:
Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'UserDefinedEvalMetric': Cannot determine Numba type of <class 'type'>

File "../python/lib/python3.8/site-packages/supervised/utils/metric.py", line 346:
    def evaluate(self, approxes, target, weight):
        <source elided>

        metric = UserDefinedEvalMetric()
        ^`

**Comment by pplonski** [comment]
Hi @Serif-Kaya,

Are you using numba jit in the custom metric? Can you paste the full code (plus example data sample) to reproduce your problem?

Thank you!

**Comment by Serif-Kaya** [comment]
@pplonski I actually resolved this issue but having trouble with train and validation set indices while using custom eval_metric. 
If I don't read train_ind it returns 
> Length of values (18) does not match length of index (20)
After using the train_ind it calculates train metric but fails at validation metric
> Length of values (2) does not match length of index (18)

`df = pd.DataFrame({'feature_1':[10, 10.0, 8, 7, 6, 5, 4, 3, 2, 1,6.0, 5.0, 4.0, 3.0, 2.0, 1.0, 10.0, 9.0, 8.0, 7.0],
                   'feature_2':[12, 8, 9, 11, 5, 7, 1.5, 4, 3, 0.5,12.0, 8.0, 9.0, 11.0, 5.0, 5.0, 1.5, 4.0, 3.0, 0.5],
                   'Group_ID':['1', '1', '1', '1', '1', '1', '1', '1', '1', '1','2', '2', '2', '2', '2', '2', '2', '2', '2', '2']
                })
def adj_rank_error(y_true, y_predicted, sample_weight=None):
    df_tmp = df.copy()
    
    train_ind = np.load(f"{mljar_path}/split_train_indices.npy")
    print(len(train_ind))

    df_tmp = df_tmp.iloc[train_ind]
    df_tmp['target'] = y_true
    df_tmp['predictions'] = y_predicted 
    
    df_tmp['rank_true'] = df_tmp.groupby("Group_ID")["target"].rank(method = "first", ascending=False)
    df_tmp['rank_predicted'] = df_tmp.groupby("Group_ID")["predictions"].rank(method = "first", ascending=False)
    
    df_tmp['rank_error'] = abs(df_tmp['rank_true'] - df_tmp['rank_predicted'])
    df_tmp['adj_rank_error'] = df_tmp['rank_error']/np.log2(1+df_tmp['rank_true'])
    
    score = np.mean(df_tmp['adj_rank_error'])
    return score`

**Comment by pplonski** [comment]
@Serif-Kaya how have you resolved the earlier problem?

Maybe try to use `train_ind` when sample is large and `validation_ind` if sample is small? You need to somehow load correct indices.

**Comment by Serif-Kaya** [comment]
I loaded the correct indices by checking the data size but sometimes y_predicted carries index and sometimes not. Working to resolve this issue now.

**Comment by somsaetang** [comment]
def NMAE(true, pred, sample_weight = None):
    
    mae = np.mean(np.abs(true - pred))
    score = mae / np.mean(np.abs(true))
    
    return score
    
This eval_metric gives 'Error for Ensemble' and 'Error for Ensemble_Stacked' when doing Compete mode.

However, if i add some code, there was no error doing Ensemble.

def NMAE(true, pred, sample_weight = None):
    if isinstance(true, pd.DataFrame):
        true = np.array(true)
    if isinstance(pred, pd.DataFrame):
        pred = np.array(pred)

    if len(pred.shape) == 2 and pred.shape[1] == 1:
        pred = pred.ravel()

    mae = np.mean(np.abs(true - pred))
    score = mae / np.mean(np.abs(true))
    
    return score

Does isinstance and ravel part is necessary when using user_defined_metric?
